### PR TITLE
PT-3240 Remove Export Redundancies

### PR DESCRIPTION
**Pull request by AlexLapena**
Removes unnecessary options within JSON and Excel export options.

### CODE REVIEW

File: components/export/ui/src/main/resources/PhenoTips/ExportPreferences.xml [hunk 1]
```diff
@@ -119,34 +121,17 @@
...
         &lt;li&gt;
-          &lt;label for="${targetFieldName}_${field.name}_code_meta"&gt;
-          &lt;input type="checkbox" name="${targetFieldName}" id="${targetFieldName}_${field.name}_code_meta" value="${field.name}_code_meta" #if($valueList.contains("${field.name}_code_meta")) checked="checked" #end /&gt;
-          ${field.translatedPrettyName} $services.localization.render('phenotips.exportPreferences.field.termAndMetaID')
-        &lt;/li&gt;
-        &lt;li&gt;
           &lt;label for="${targetFieldName}_${field.name}_meta"&gt;
           &lt;input type="checkbox" name="${targetFieldName}" id="${targetFieldName}_${field.name}_meta" value="${field.name}_meta" #if($valueList.contains("${field.name}_meta") || $valueListisEmpty) checked="checked" #end /&gt;
-          ${field.translatedPrettyName} $services.localization.render('phenotips.exportPreferences.field.termAndMetaNames')
+          ${field.translatedPrettyName} $services.localization.render('phenotips.exportPreferences.field.qualifiers')
```
Comment (sdumitriu) [inline components/export/ui/src/main/resources/PhenoTips/ExportPreferences.xml]: This kind of translation chunking doesn't work, not all languages use the same topic. The full sentence should be translated.

**Comment by AlexLapena** [comment]
@sdumitriu Code review changes added and rebased

**Comment by sashaandjic** [comment]
Looks much better now without redundant options. However, both export dialogs (excel and JSON) are still quite ugly. Many fields that should be grouped together are showin somewhere else (e.g. Clinical, Final diagnosis and Additional comments should be displayed together). We should make both dialogs ui wise consistent with push dialog fields selection.

File: components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java [hunk 2]
```diff
@@ -86,28 +86,15 @@ public void phenotypeSetup(Set<String> enabledFields) throws Exception
...
-            { "id", "positive" },
-            { "name", "id", "positive" },
-            { "meta.id", "name", "positive" },
-            { "meta.name", "name", "positive" },
-            { "name", "negative" },
-            { "id", "negative" },
-            { "id", "name", "negative" },
-            { "type" }
+                { "name", "id", "positive", "negative" },
+                { "meta.name", "name", "positive", "negative" },
```
Comment (sdumitriu) [inline components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java]: `meta.id` should still be used, don't remove it.

File: components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java [hunk 3]
```diff
@@ -86,28 +86,15 @@ public void phenotypeSetup(Set<String> enabledFields) throws Exception
...
+                { "type" }
```
Comment (sdumitriu) [inline components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java]: This one has no equivalent trigger in `fieldIds`, so it's dead.

File: components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java [hunk 4]
```diff
@@ -182,7 +168,7 @@ public DataSection phenotypeBody(Patient patient) throws Exception
         for (Feature feature : sortedFeatures) {
             x = 0;
 
-            if (bothTypes && lastStatus != feature.isPresent()) {
+            if (lastStatus != feature.isPresent()) {
```
Comment (sdumitriu) [inline components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java]: Bug: if the patient only has negative phenotypes, no label will be printed.

File: components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java [hunk 5]
```diff
@@ -207,48 +191,28 @@ public DataSection phenotypeBody(Patient patient) throws Exception
...
-                        DataCell cell = new DataCell(meta.getName(), mX + offset, y);
-                        section.addCell(cell);
-                        offset += 1;
-                    }
-                    if (present.contains("meta.id")) {
-                        DataCell cell = new DataCell(meta.getId(), mX + offset, y);
-                        section.addCell(cell);
-                        offset += 1;
-                    }
+                    section.addCell(new DataCell(meta.getName(), mX + offset, y));
```
Comment (sdumitriu) [inline components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java]: Now offset is always `0`, so it's useless.

File: components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java [hunk 6]
```diff
@@ -207,48 +191,28 @@ public DataSection phenotypeBody(Patient patient) throws Exception
...
                 Boolean metaPresent = !featureMetadatum.isEmpty();
                 int offset = 0;
                 for (FeatureMetadatum meta : featureMetadatum) {
                     offset = 0;
-                    if (present.contains("meta.name")) {
-                        DataCell cell = new DataCell(meta.getName(), mX + offset, y);
-                        section.addCell(cell);
-                        offset += 1;
-                    }
-                    if (present.contains("meta.id")) {
```
Comment (sdumitriu) [inline components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java]: The id shouldn't be removed.

File: components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java [hunk 7]
```diff
@@ -274,27 +238,28 @@ public void variantsSetup(Set<String> enabledFields) throws Exception
         String sectionName = "variants";
         String[] fieldIds =
             { "variants", "variants_protein", "variants_transcript", "variants_dbsnp", "variants_zygosity",
-            "variants_effect", "variants_interpretation", "variants_inheritance", "variants_evidence",
-            "variants_segregation", "variants_sanger", "variants_coordinates" };
+                "variants_effect", "variants_interpretation", "variants_inheritance", "variants_evidence",
```
Comment (sdumitriu) [inline components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java]: These changes are irrelevant for this issue, so they shouldn't be included in the pull request.

File: components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java [hunk 8]
```diff
@@ -1116,25 +1070,16 @@ public DataSection prenatalPhenotypeBody(Patient patient) throws Exception
...
-                x++;
-            }
-            if (present.contains("id")) {
-                DataCell cell = new DataCell(feature.getId(), x, y, StyleOption.FEATURE_SEPARATOR);
-                section.addCell(cell);
-                x++;
-            }
+            DataCell cell = new DataCell(feature.getName(), ++x, y, StyleOption.FEATURE_SEPARATOR);
+            section.addCell(cell);
+            section.addCell(new DataCell(feature.getId(), ++x, y, StyleOption.FEATURE_SEPARATOR));
```
Comment (sdumitriu) [inline components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java]: These two cells are created the same way, so why two different ways? The name cell should also be added directly, without storing it in a temporary variable.

File: components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java [hunk 9]
```diff
@@ -1147,48 +1092,30 @@ public DataSection prenatalPhenotypeBody(Patient patient) throws Exception
     public DataSection disordersHeaders(Set<String> enabledFields) throws Exception
     {
         String sectionName = "disorders";
+        Set<String> present = new HashSet<>();
+        present.add("omim_id"); //Accepts one option only
```
Comment (sdumitriu) [inline components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java]: Use `Set<String> present = Collections.singleton("omim_id");`

File: components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java [hunk 10]
```diff
@@ -1205,32 +1132,21 @@ public DataSection disordersBody(Patient patient)
...
-                x++;
-            }
+            DataCell cell = new DataCell("", x, y);
+            bodySection.addCell(cell);
+            x++;
+            bodySection.addCell(new DataCell("", x, y));
+            x++;
         }
         /* Notes export */
         if (present.contains("notes")) {
```
Comment (sdumitriu) [inline components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java]: This `if` is now redundant, and actually leaving it in means that notes will not be displayed anymore.

File: components/export/ui/src/main/resources/PhenoTips/ExportPreferences.xml [hunk 11]
```diff
@@ -109,6 +110,8 @@
   {{html clean="false"}}
     #set($valueListisEmpty = ($valueList.size() == 0))
     #set($xclass = $xwiki.getDocument($xclassName).getxWikiClass())
+    #set ($discard = $ignored.fields.add('negative_phenotype'))
+    #set ($discard = $ignored.fields.add('negative_prenatal_phenotype'))
```
Comment (sdumitriu) [inline components/export/ui/src/main/resources/PhenoTips/ExportPreferences.xml]: Why? Aren't these already included in `$ignored.fields`?

File: components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java (hunk 7 above)
Comment (AlexLapena) [inline components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java]: I used my Phenotips formatting tool before the PR, should I have only used it on the section of code I was working on?

Comment (sdumitriu) [inline components/export/api/src/main/java/org/phenotips/export/internal/DataToCellConverter.java]: It's OK to use it everywhere, but you can skip these changes when doing a specific pull request.
